cam/rawsegmenter.go

```diff
@@ -0,0 +1,254 @@
+	resetTimestampsCStr := C.CString("1")
+	strftimeCStr := C.CString("1")
+	defer func() {
+		C.free(unsafe.Pointer(segmentLengthCStr))
+		C.free(unsafe.Pointer(segmentFormatCStr))
+		C.free(unsafe.Pointer(resetTimestampsCStr))
+		C.free(unsafe.Pointer(strftimeCStr))
+	}()
+
+	// Segment options are passed as opts to avformat_write_header. This only needs
+	// to be done once, and not for every segment file.
+	var opts *C.AVDictionary
+	defer C.av_dict_free(&opts)
+	ret = C.av_dict_set(&opts, C.CString("segment_time"), segmentLengthCStr, 0)
+	if ret < 0 {
+		return fmt.Errorf("failed to set segment_time: %s", ffmpegError(ret))
+	}
+	ret = C.av_dict_set(&opts, C.CString("segment_format"), segmentFormatCStr, 0)
+	if ret < 0 {
+		return fmt.Errorf("failed to set segment_format: %s", ffmpegError(ret))
+	}
+	ret = C.av_dict_set(&opts, C.CString("reset_timestamps"), resetTimestampsCStr, 0)
+	if ret < 0 {
+		return fmt.Errorf("failed to set reset_timestamps: %s", ffmpegError(ret))
+	}
+	ret = C.av_dict_set(&opts, C.CString("strftime"), strftimeCStr, 0)
+	if ret < 0 {
+		return fmt.Errorf("failed to set strftime: %s", ffmpegError(ret))
+	}
+
+	// Open the output file for writing
+	ret = C.avformat_write_header(fmtCtx, &opts)
+	if ret < 0 {
+		return fmt.Errorf("failed to write header: %s", ffmpegError(ret))
+	}
+
+	// Need to reset stream time base to 1/1000
+	stream.time_base.num = 1
+	stream.time_base.den = 1000
+
+	return nil
+}
+
+func (rs *RawSegmenter) WritePacket(payload []byte, pts int64, isIDR bool) error {
+	// Stuff the bytes payload and timestamps into an AV Packet.
+	avpkt := C.av_packet_alloc()
+	if avpkt == nil {
+		return errors.New("failed to allocate AVPacket")
+	}
+	defer C.av_packet_free(&avpkt)
+	// TODO(seanp): should i use (*C.uint8_t)(unsafe.Pointer(&payload[0])) or C.CBytes(payload) instead?
+	avpkt.data = (*C.uint8_t)(C.av_malloc(C.size_t(len(payload))))
+	if avpkt.data == nil {
+		C.av_packet_free(&avpkt)
+		return errors.New("failed to allocate AVPacket data")
+	}
+	// Copy the contents of the Go byte slice into the allocated memory
+	C.memcpy(unsafe.Pointer(avpkt.data), unsafe.Pointer(&payload[0]), C.size_t(len(payload)))
+	avpkt.size = C.int(len(payload))
+	if avpkt.size == 0 {
+		return errors.New("empty packet")
+	}
+	if avpkt.data == nil {
+		return errors.New("nil packet data")
+	}
+	avpkt.pts = C.int64_t(pts)
+	avpkt.dts = C.int64_t(pts)
+	// Set the keyframe flag if this is an IDR frame. This is needed to make sure the
+	// muxer knows it is a keyframe and is safe to start a new segment.
+	if isIDR {
+		avpkt.flags |= C.AV_PKT_FLAG_KEY
+	}
+	// Write the packet to the output file.
+	ret := C.av_interleaved_write_frame(rs.outCtx, avpkt)
+	if ret < 0 {
+		return fmt.Errorf("failed to write frame: %s", ffmpegError(ret))
+	}
+	return nil
+}
+
+// Close closes the segmenter and writes the trailer to prevent corruption
+// when exiting early in the middle of a segment.
+func (rs *RawSegmenter) Close() {
```

Comment:
[seanp] This handles the case when the program early exits and we need to terminate a clip before it is finished.